File: zoouniak/Programmers/도둑질.java
package Programmers;

public class 도둑질 {
    class Solution {
        public int solution(int[] money) {
            int[] answers = new int[money.length];
            int[] noFirstHouse = new int[money.length];
            answers[0] = money[0];
            noFirstHouse[0] = 0;
            answers[1] = Math.max(money[0], money[1]);
            noFirstHouse[1] = money[1];
            for (int i = 2; i < money.length; i++) {
                if (i == money.length - 1) {
                    //마지막 집이 선택되면 첫번째 집은 선택되지 않아야함
                    answers[i] = Math.max(noFirstHouse[i - 2] + money[i], answers[i - 1]);
                } else {
                    answers[i] = Math.max(answers[i - 2] + money[i], answers[i - 1]);
                    noFirstHouse[i] = Math.max(noFirstHouse[i - 2] + money[i], noFirstHouse[i - 1]); //첫번째 집 선택안하는 경우
                }
            }
            return answers[money.length - 1];
        }
    }
}

File: dnaeuuon/Programmers/도둑질.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>

using namespace std;
int dp1[1000001];
int dp2[1000001];

int solution(vector<int> money) {
    int answer = 0;
    int n = money.size() - 1;
    
    dp1[0] = money[0];
    dp1[1] = dp1[0];
    dp2[0] = 0;
    dp2[1] = money[1];
    
    for(int i = 2; i < n; i++) {
        dp1[i] = max(dp1[i-2] + money[i], dp1[i-1]);
    }
    for(int i = 2; i <= n; i++) {
        dp2[i] = max(dp2[i-2] + money[i], dp2[i-1]);
    }
    answer = max(dp1[n-1], dp2[n]);
    return answer;
}
